### PR TITLE
JVM IR: Don't generate nullability assertions for parameters of directly invoked lambdas

### PR DESCRIPTION
**Pull request by scaventz**
issue: [KT-46615](https://youtrack.jetbrains.com/issue/KT-46615)

### CODE REVIEW

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt [hunk 1]
```diff
@@ -478,6 +482,10 @@ class ExpressionCodegen(
             val parameterType = callable.valueParameterTypes[i]
             require(arg != null) { "Null argument in ExpressionCodegen for parameter ${irParameter.render()}" }
             callGenerator.genValueAndPut(irParameter, arg, parameterType, this, data)
+
+            if (expression.symbol.owner.origin == IrDeclarationOrigin.LOCAL_FUNCTION_FOR_LAMBDA && arg is IrConst<*>) {
```
Comment (dnpetrov) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt:486]: `IrConst<*>` expression is just a constant literal. It doesn't imply anything about given call expression being a result of direct invoke optimization for a lambda.

**Comment by dnpetrov** [comment]
Directly invoked lambdas don't need parameter nullability assertions, because corresponding methods are not a part of the public ABI. They have a single call site that is type-checked by Kotlin front-end, so corresponding run-time check in the method itself can be skipped. Arguments themselves, constant or not, are not related to this optimization (they are checked at the call site if required, e.g., if these are potentially nullable platform values used where a value of non-null type is expected).

Proper way to implement this optimization would be to track directly invoked lambdas somehow and check it during parameter nullability assertions generation.

**Comment by scaventz** [comment]
Thanks for the background information, I'm still learning 'JVM IR', please pardon me if I understand this wrong.

The following example makes me unsure, should it be considered to be a directly invoked lambda? If it is, it seems that a `NullPointerException` will be produced at runtime if we remove the parameter nullability assertions

```kotlin
// FILE: A.java
class A {
    public static String nullFromJavaAtRuntime = null;
}
// FILE: main.kt
fun main() {
    { s: String -> println(s.length) }(A.nullFromJavaAtRuntime)
}
```

**Comment by dnpetrov** [comment]
Kotlin generates two kinds of null checks:

(1) Argument null checks at call site, where a possibly null plaform value is used where a non-null value is expected.  Such checks look like `INVOKESTATIC kotlin/jvm/internal/Intrinsics.checkNotNullExpressionValue (Ljava/lang/Object;Ljava/lang/String;)V` in the bytecode. Such null check is effectively an "implicit cast" of a platform type value to a non-null type. These checks would stay there regardless of what we are actually calling. We just prefer to fail fast in such case.
Your example fails with NPE in method `main` with a message like `A.nullFromJavaAtRuntime must not be null`. It is an argument null check. 

(2) Parameter null checks at declaration site (in the beginning of the corresponding JVM method). When you look into the bytecode generated by Kotlin compiler, these checks look like `INVOKESTATIC kotlin/jvm/internal/Intrinsics.checkNotNullParameter (Ljava/lang/Object;Ljava/lang/String;)V`. Parameter null checks are there mostly to preserve null safety when calling Kotlin code from some other language, e.g., from Java. Such checks are required for methods that are a part of public ABI. 
Methods for directly invoked lambdas are not a part of public ABI. Such lambda is invoked immediately and doesn't even exist as an object that might escape to, say, some Java code. So, methods for directly invoked lambdas don't need parameter null checks.

**Comment by dnpetrov** [comment]
LGTM. I'll run tests and push.

**Comment by dnpetrov** [comment]
Pushed ( https://github.com/JetBrains/kotlin/commit/8317daa00ba55c07e44e67f3e3e010616c981967 ).
Thank you for your contribution!

**Pull request closed without merging**